### ISSUE TITLE
Wildcards in directory

### ISSUE THREAD

**Issue by blakemorgan** [dependabot/dependabot-core#2178]
Our repo doesn't have a `pom.xml` file at `/` but it does in multiple subdirectories (so `/*/`). Dependabot didn't recognize the wildcard in the directory name when I put that in the config file. Is it possible to get wildcards recognized in the directory value of the config file?

**Comment by alannt777** [comment]
This is extremely important feature to have. It's frustrating that I can't use wildcards for large codebases

**Comment by alexf101** [comment]
I wrote a script that auto-generates Dependabot configs using globs 😅 

Please feel free to use my awful hacks however you like until Dependabot/Github implement this feature properly.

https://gist.github.com/alexf101/b65cbfe7c5a61df7d925589a71d200cf

**Comment by jauderho** [comment]
+1 from me.

I have a dockerfiles repo with multiple directories representing different images being built. I was dismayed to find out that dependabot.yml only allows specifying one directory.

Even if we are somehow able to do the following would be a start

> directory: |
>     /dir1
>     /dir2
```

**Comment by michaelmerrill** [comment]
Any update on where this is on the roadmap?

**Comment by blakemorgan** [comment]
For our teams, the Terraform support with Dependabot greatly increased the need for this feature. We have local modules storing common infrastructure for the environments, then a separate folder for each environment. Sometimes, because of weird dependencies, we have to have two folders for each environment. That's upwards of 10 folders for each project. It would be nicer to have one entry in the config file instead of 10.

**Comment by bjaanes** [comment]
If someone would like a javascript hack as a workaround:
- handlebar template for dependabot file
- "fast-glob" all the different types you need and generate a new dependabot file:
```js
fb.sync([
        `${repoRoot}/**/package.json`,
        `${repoRoot}/**/go.mod`,
        `${repoRoot}/**/pom.xml`,
        `${repoRoot}/**/Dockerfile`,
        // Stuff to ignore:
        `!${repoRoot}/**/node_modules`,
        `!${repoRoot}/**/vendor`])
        .map(mapItToAStructureYouWantToSendToHandleBar)
```

For monorepos, it is death trying to maintain the dependabot file by hand.

**Comment by tmatilai** [comment]
Generating the configuration for each and every Terraform module in a repository worked fine, until:

> `The property '#/updates' had more items than the allowed 200`

No idea how to work around this now. 🙁

**Comment by highb** [comment]
@greysteil Any estimates on when this may be implemented in Dependabot? Managing our Monorepo with one definition for every requirements.txt and package.json is not sustainable for us.

**Comment by greysteil** [comment]
Afraid not - I don't work on Dependabot directly anymore and haven't for the last couple of years. Cc @exvuma.

**Comment by exvuma** [comment]
It's not on our roadmap right now tbh. It's a large engineering undertaking but is top of mind on how to address this sort of configurability issue for monorepos. I am tracking this internally. As we nail down our planning I will update this ticket. In the meantime, you can see the features we commit to shipping on [our public roadmap](https://github.com/orgs/github/projects/4247).

**Comment by DennisDyallo** [comment]
Really want this feature. We have multiple csproj files in subdirectories and now we are forced to list them all by path in the dependabot.yml, instead of just doing /rootDir/**.csproj

**Comment by gaby** [comment]
👍🏻 to get this feature added

**Comment by prapurnakumariTR** [comment]
Is this feature available?

**Comment by arielweinberger** [comment]
Any news about this? It's been open since 2019. Can we at least get a confirmation of whether this is something you plan to support or not? This could be a deal breaker for us, between Dependabot and Renovate.

**Comment by gtirloni** [comment]
> It's not on our roadmap right now tbh. It's a large engineering undertaking but is top of mind on how to address this sort of configurability issue for monorepos. I am tracking this internally. As we nail down our planning I will update this ticket. In the meantime, you can see the features we commit to shipping on [our public roadmap](https://github.com/orgs/github/projects/4247).

It's not even in the Future column for dependabot. Is there any interest in adding this feature in 2022?

**Comment by cyberveseli** [comment]
Vote for this feature 👍

**Comment by rutalreja-deloitte** [comment]
For TF files, can be modified for any extension
A little late to the party, but finally here

One Liner that I run in a git actions pipeline (remember to give pipeline push permissions) from root of repo
`
if [[ "$(find . -name '*.tf' -not -path '*.terraform*' -printf '%h\n' | sort -u | wc -l)" -lt "200" ]]; then yq -i 'del(.updates.[])' .github/dependabot.yml && for DIR in $(find . -name '*.tf' -not -path '*.terraform*' -printf '%h\n' | sort -u); do dir=${DIR:1} yq -i '."updates"+=[{"package-ecosystem":"terraform" | . style="double","directory":env(dir) | . style="double","schedule":{"interval":"daily" | . style="double"}}]' .github/dependabot.yml; done;fi
`

A little explaination:
- Firstly it uses `yq`, if you cant deal with the dependency, enjoy doing it with `jq` yourself
- `find . -name '*.tf' -not -path '*.terraform*' -printf '%h\n' | sort -u` finds the names of the folders that contain `.tf` files apart from inside the `.terraform` folder
- The first if condition `if [[  ... -lt "200" ]]` checks you dont have more than 200 folders, as ![@tmatilai](https://github.com/dependabot/dependabot-core/issues/2178#issuecomment-1029740098) mentioned above. If you do good luck.
- `yq -i 'del(.updates.[])' .github/dependabot.yml` this clears your dependabot updates yaml array to remake it in the next line.
   - If you have other things you dont want to delete you will have to figure out what is old and what is new if you are running in a pipeline, a head start for that would be something like `diff <(command) <(command)` to compare
- Last `for DIR in $(find . -name '*.tf' -not -path '*.terraform*' -printf '%h\n' | sort -u); do dir=${DIR:1} yq -i '."updates"+=[{"package-ecosystem":"terraform" | . style="double","directory":env(dir) | . style="double","schedule":{"interval":"daily" | . style="double"}}]' .github/dependabot.yml` loops through, creates the structure and fills in the directory

**Comment by Snailedlt** [comment]
Would be neat if directory could take a list or an array of paths too... 
maybe something like this?
```yml
  - package-ecosystem: "composer"
    directory: 
      - "/app"
      - "/"
    schedule:
      interval: "daily"
```

**Comment by fullstackzach** [comment]
I would like to throw my hat in the ring to request this feature too. We use a monorepo that has several package.json files in different folders in the project.

All I am looking to do is create the dependabot.yml in order to add a prefix to the commit message, but that means I would have to list all of the locations of package.json files explicitly.

Interestingly without a dependabot.yml, dependabot security updates appears to be able to automatically detect where these package.json files exist.

**Comment by Makeshift** [comment]
Since this has been going on for a while and probably isn't going to be added soon, I created a little workaround in the form of an action that generates `dependabot.yml` based on globs.

[Example template](https://github.com/Makeshift/generate-dependabot-glob-action/blob/master/.github/dependabot.template.yml) -> [Example output](https://github.com/Makeshift/generate-dependabot-glob-action/blob/master/.github/dependabot.yml)

https://github.com/marketplace/actions/generate-dependabot-glob

It's exceptionally simple and was created in a rush as opposed to with good practice, but hopefully is enough for simple projects/monorepos or can be used as a base to build your own generator from. It's certainly a lot nicer to have a 20 line file that generates into a 300 line one automatically!

**Comment by Dreamsorcerer** [comment]
Thanks to #6255, I've just found a partial workaround for Python projects.

You can add a top-level requirements file which includes all your other requirements files (with `-r sub/requirements.txt`). It will then check them all recursively. Example: https://github.com/aio-libs/aiohttp-demos/pull/540/files

Obviously doesn't resolve the discoverability issue (you still need to hardcode every requirements file), but it means you'll now get 1 PR to update a dependency across all requirements files, instead of a separate PR for each file.

**Comment by gaby** [comment]
4 years later, are there any updates on this?

**Comment by kashifest** [comment]
This is really desirable now, we have for example four different go modules in sub directories in https://github.com/metal3-io/cluster-api-provider-metal3/ , we would really love to put the directories in an array and possibly also hope the updates of one dependancy goes into one PR no matter if it is updating multiple gomodules or not.

**Comment by arslanbekov** [comment]
After added the ability to groups feature, there is very needed support for specifying wildcard in the directory.

**Comment by Dreamsorcerer** [comment]
> Thanks to #6255, I've just found a partial workaround for Python projects.

It appears this no longer works. I've just had 8 PRs open for the same dependency. And every PR fails because the versions obviously conflict with each other.

**Comment by Dreamsorcerer** [comment]
Oh, I think I just realised it was caused by a security update. So, security updates came in one at a time, which all failed, then the regular version update did the normal batch update, and made the other PRs redundant.

**Comment by dgokcin** [comment]
has anyone tried the described options in [github blog](https://github.blog/2023-08-24-a-faster-way-to-manage-version-updates-with-dependabot/)? looks promising.

**Comment by ilmax** [comment]
@dgokcin I am using it and it somehow works, what I can't do with grouped updates is to ignore a dependency version that's part of the group. The command is there in the dependabot comment but it looks like is not doing anything.

**Comment by rdhar** [comment]
> @dgokcin I am using it and it somehow works, what I can't do with grouped updates is to ignore a dependency version that's part of the group. The command is there in the dependabot comment but it looks like is not doing anything.

Mind sharing a link to an example where you've somehow got wildcard-directory-support working within dependabot.yml?

**Comment by dgokcin** [comment]
> @dgokcin I am using it and it somehow works, what I can't do with grouped updates is to ignore a dependency version that's part of the group. The command is there in the dependabot comment but it looks like is not doing anything.

@ilmax thank you for sharing what has worked and what has not worked for you. as @rdhar said an example would be great.

**Comment by ilmax** [comment]
> > @dgokcin I am using it and it somehow works, what I can't do with grouped updates is to ignore a dependency version that's part of the group. The command is there in the dependabot comment but it looks like is not doing anything.
> 
> Mind sharing a link to an example where you've somehow got wildcard-directory-support working within dependabot.yml?

Wait, I may have misunderstood a bit. the linked post is about grouped updates and that's what I was referring to in my comment. 
This issue on the contrary is about having dependabot support some sort of search within the project folders. This second part is still not working, but if you want an example of the first one, I can post it here

**Comment by cprice404** [comment]
This has turned into a blocker for us for a few of our repos. The repos have 10+ subdirectories that have example code in them using different combinations of dependencies, and listing each one as a `package-ecosystem` means that we regularly have a dozen dependabot PRs open that are all consuming Actions runners. It is looking like we will have to disable dependabot because that many PRs/actions are not sustainable for us, and we expect the number of directories to grow.

**Comment by jan-xyz** [comment]
You could also change the base branch of the PRs to one of the 15 other PRs and merge them all together before merging to main.

**Comment by cpswan** [comment]
To do what @jan-xyz  suggests I have a little script - [rollup.sh](https://gist.github.com/cpswan/1e61f12b4a83355a7560d96cb1cdf726)

More explanation in the accompanying blog post [Roll up, rollup, get your Dependabot PRs together here](https://blog.thestateofme.com/2023/03/09/roll-up-rollup-get-your-dependabot-prs-together-here/)

**Comment by ascopes** [comment]
Any update on this? I have a project for a Maven plugin that has integration tests taking the shape of multiple maven projects in `src/it/*/pom.xml` that would benefit from this feature.

**Comment by reggermont** [comment]
> Any update on this? I have a project for a Maven plugin that has integration tests taking the shape of multiple maven projects in `src/it/*/pom.xml` that would benefit from this feature.

The issue is 5 years old, pretty sure you can forget about it

**Comment by burik666** [comment]
we need it

**Comment by gioce90** [comment]
We absolutely need this feature. Anything yet?

Maybe replacing the `directory` field with a `directories`:

```
updates:
  - package-ecosystem: "nuget"
    directories:
    - "/App/First" 
    - "/App/Second"
    open-pull-requests-limit: 15
    # other settings...
```

**Comment by Dreamsorcerer** [comment]
Possibly the same status as https://github.com/dependabot/dependabot-core/issues/4364#issuecomment-1887660591

**Comment by PrimeDominus** [comment]
Woah! How is this not a standard feature already???

**Comment by jonasfroeller** [comment]
This issue is from 2019. How is this not implemented yet?

**Comment by adamlui** [comment]
I'd like it too...

**Comment by Parth-1612** [comment]
same

**Comment by Mayeu** [comment]
If any of you really need this feature then vote with your feet and migrate to another dependency manager, like Renovate (https://github.com/renovatebot/renovate) it's almost 0 config with autodetection of any number of files needed to be updated.

(I'm not affiliated with this tool, I'm just a happy user since I took the 15 minutes needed to migrate.)

**Comment by gioce90** [comment]
> If any of you really need this feature then vote with your feet and migrate to another dependency manager, like Renovate (https://github.com/renovatebot/renovate) it's almost 0 config with autodetection of any number of files needed to be updated.
> 
> (I'm not affiliated with this tool, I'm just a happy user since I took the 15 minutes needed to migrate.)

I came to the same conclusion a few weeks ago (I'm not affiliated too).
Renovate is a really great alternative, especially outside of GitHub (I was looking for something that worked for Azure DevOps).

**Comment by Snailedlt** [comment]
> > If any of you really need this feature then vote with your feet and migrate to another dependency manager, like Renovate (https://github.com/renovatebot/renovate) it's almost 0 config with autodetection of any number of files needed to be updated.
> > 
> > (I'm not affiliated with this tool, I'm just a happy user since I took the 15 minutes needed to migrate.)
> 
> I came to the same conclusion a few weeks ago (I'm not affiliated too).
> Renovate is a really great alternative, especially outside of GitHub (I was looking for something that worked for Azure DevOps).

Agreed, though I personally prefer dependabot due to its tight integration on GitHub.

Just FYI, dependabot also works on Azure DevOps :)

**Comment by robinshin** [comment]
Same issue here

**Comment by abdulapopoola** [comment]
The team has started working on this and hoping we have some releases soon. Tagging @jurre , @jakecoffman , and @Nishnha

**Comment by ascopes** [comment]
![h9lvh3l5uwh61](https://github.com/dependabot/dependabot-core/assets/73482956/f4979fe9-8be0-450c-b7eb-257543cc0d21)

This sounds great. Look forward to trying this out once it is ready. Thank you!

**Comment by moritzschmitz-oviva** [comment]
Great, now I need to rewrite all the tooling we made to glue this together :trollface:

**Comment by damianh** [comment]
🙏🏻 Please make it support wild cards _and_ multiple directory entries 🙏🏻

**Comment by abdulapopoola** [comment]
That's the plan; tagging @jurre 😄

**Comment by abdulapopoola** [comment]
Hopefully some time in April :), if you are interested in an early beta, that might be possible. cc: @carlincherry

**Comment by RDhar** [comment]
> Hopefully some time in April :), if you are interested in an early beta, that might be possible. cc: @carlincherry

Ooh, I'd be happy apply for early-beta-feedback as well (my use-case is with `terraform` package manager, which is a [surprisingly popular](https://github.com/dependabot/dependabot-core/issues/2178#issuecomment-861882801) "niche").

**Comment by danielkubat** [comment]
> Hopefully some time in April :), if you are interested in an early beta, that might be possible. cc: @carlincherry 

any estimation, when this feature will be ported to GitHub Enterprise?

**Comment by kaefferlein** [comment]
> Hopefully some time in April :), if you are interested in an early beta, that might be possible. cc: @carlincherry

absolutely interested in early beta :)

**Comment by gonzalemario** [comment]
Are we waiting Halley comet to show up?

**Comment by ascopes** [comment]
Will the beta be available on GitHub Community? How does that usually work?

**Comment by damianh** [comment]
> Hopefully some time in April :), if you are interested in an early beta, that might be possible. cc: @carlincherry

✋🏻 Count me in for early beta 😄

**Comment by dgokcin** [comment]
also interested in any sort of beta ✋

**Comment by neilime** [comment]
Hi, also interested of beta :+1:

**Comment by ascopes** [comment]
@adamlui I don't think the comments referring to version naming here are going to impact how the team manages their release schedule. If it is an urgent issue then more people need to upvote the GitHub issue itself.

That aside, people registering interest in using the release candidate versions means the feature gets more usage exposure sooner, allowing identification of bugs sooner.

Urgent or not, it needs to be tested properly

**Comment by lucacome** [comment]
I noticed that dependabot started using a new format for the title of grouped updates, like
```
Bump the actions group across 1 directory with 5 updates
```

Does it mean that support for multiple directories is almost here? 😄

**Comment by burik666** [comment]
@lucacome [Dependabot grouped security updates generally available](https://github.blog/changelog/2024-03-28-dependabot-grouped-security-updates-generally-available/)

**Comment by adamlui** [comment]
> @adamlui I don't think the comments referring to version naming here are going to impact how the team manages their release schedule. If it is an urgent issue then more people need to upvote the GitHub issue itself.

@ascopes you can think what you want, doesn't change the fact if a million ppl complain about some bug in my software I will absolutely treat it less urgent if everyone keeps expressing interest in trialing some "beta" (this is merely common human psychology, your demand ppl upvote issue does not affect it)

> That aside, people registering interest in using the release candidate versions means the feature gets more usage exposure sooner, allowing identification of bugs sooner.

False, again, if something is touted as a beta feature, there is less incentive to treat it as an issue instsead (features are bonuses, issues are not)

> Urgent or not, it needs to be tested properly. I'd much rather have something that works properly and is battle tested than something that raises erroneous MRs on my projects.

This is a poor justification for calling something beta (you can test patches without calling them beta)

**Comment by adamlui** [comment]
You can get mad but if you reflect on how utterly simple parsing a YAML containing an array of dirs should be yet it's somehow still not implemented, everyone hyping it up as some excusable "beta feature" instead of the simple bug fix it should've been cannot possibly be helping the situation

**Comment by Dreamsorcerer** [comment]
Nor is complaining if it's so easy for you to implement... code's right there under the 'Code' tab.

**Comment by adamlui** [comment]
Knew someone would say something like you. GitHub programmers are paid six figures, we are not, stop trying to excuse laziness (it's been 5 years literally wake up)

**Comment by adamlui** [comment]
(it is very easy but don't ever code for free for software that is maintained by paid coders unless your time is utterly valueless)

**Comment by abdulapopoola** [comment]
[Multi-dir configuration is now in public beta!](https://github.blog/changelog/2024-04-29-dependabot-multi-directory-configuration-public-beta-now-available/); open to hearing your thoughts as you try out the beta!!

Tagging @carlincherry

**Comment by mhemmings** [comment]
@abdulapopoola @carlincherry Amazing, thanks!

How does this work with Github Actions, since they have slightly different behaviour with regards to `directory`? Does it fix the ability to update composite actions? (See https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#github-actions)

i.e. I'm hoping the following should now work:

```yaml
version: 2
updates:
  - package-ecosystem: "github-actions"
    directories:
      - "./github/workflows"
      - "./github/actions/composite-a"
      - "./github/actions/composite-b"
      - "./github/actions/composite-c"
    schedule:
      interval: "weekly"

```

**Comment by blu3mania** [comment]
> [Multi-dir configuration is now in public beta!](https://github.blog/changelog/2024-04-29-dependabot-multi-directory-configuration-public-beta-now-available/); open to hearing your thoughts as you try out the beta!!
> 
> Tagging @carlincherry

Question, is the beta available to Enterprise Cloud repos?

**Comment by lucacome** [comment]
I was trying this out and it doesn't seem to work.

Well, it did group all the different directories in the dependabot tab, but then when there was a dependency update it only updated the first `go.mod` file in the `directories` list and not the second one.

**Comment by cpswan** [comment]
I'm seeing similar problems to @lucacome :(

Now rather than getting 4 PRs in one day I'm going to see 4 PRs over 4 days (for the same dependency change), which isn't an improvement on toil, and makes things worse for timely updates.

What's supposed to happen here is all 4 PRs rolled into one.

It makes me wonder if anybody actually tested the 'beta'?

**Comment by carlincherry** [comment]
> How does this work with Github Actions, since they have slightly different behaviour with regards to directory? Does it fix the ability to update composite actions? (See https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#github-actions)

@mhemmings I would expect this to work with composite actions. Did you experience different behavior?

**Comment by carlincherry** [comment]
> Question, is the beta available to Enterprise Cloud repos?

@blu3mania yes, this beta should be available to Enterprise Cloud repos!

**Comment by carlincherry** [comment]
> I was trying this out and it doesn't seem to work.
> 
> Well, it did group all the different directories in the dependabot tab, but then when there was a dependency update it only updated the first go.mod file in the directories list and not the second one.

@lucacome thank you for raising this to the team; this is unexpected behavior and we've created a [bug](https://github.com/dependabot/dependabot-core/issues/9664) which our team will prioritize fixing soon. @cpswan is your bug happening with `go` as well?

cc @abdulapopoola

**Comment by carlincherry** [comment]
@cpswan is your bug happening with `go` as well? Do you have [grouping](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#groups) enabled?

**Comment by adamlui** [comment]
> > I was trying this out and it doesn't seem to work.
> > Well, it did group all the different directories in the dependabot tab, but then when there was a dependency update it only updated the first go.mod file in the directories list and not the second one.
> 
> @lucacome thank you for raising this to the team; this is unexpected behavior and we've created a [bug](https://github.com/dependabot/dependabot-core/issues/9664) which our team will prioritize fixing soon.
> 
> cc @abdulapopoola

@carlincherry since you appear to have the higher authority to prioritize bugs, can this one be added https://github.com/dependabot/dependabot-core/issues/9071 as it is quite stale and cost me endless hours of extra manual labor that should not be necessary (dependabot is supposed to automate, not create more work than just typing `npm update`)

**Comment by cpswan** [comment]
> @cpswan do you have [grouping](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#groups) enabled? cc @abdulapopoola

@carlincherry  thanks for the pointer. Adding a group seems to get the behaviour I was expecting. It might be worth being more explicit about that in the documentation. Essentially `directories:` doesn't work (in a meaningful way) without an accompanying `groups:`.

Here's my working YAML fragment:

```yaml
  - package-ecosystem: "docker"
    directories:
      - "/some_dockerfiles/"
      - "/another_place/with_dockerfiles/"
    schedule:
      interval: "daily"
    groups:
      docker:
        patterns:
          - "*"
 ```

**Comment by ffried** [comment]
> > @cpswan do you have [grouping](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#groups) enabled? cc @abdulapopoola
> 
> @carlincherry thanks for the pointer. Adding a group seems to get the behaviour I was expecting. It might be worth being more explicit about that in the documentation. Essentially `directories:` doesn't work (in a meaningful way) without an accompanying `groups:`.


Wouldn't that mean, that you no longer get multiple PRs for different dependencies?
E.g. if a repositories contains three directories, each with an `npm` project, and each project having e.g. `semver` and `express` as dependencies. When using `groups` as shown above, this would mean, that there's one big PR containing updates to both `semver` and `express` for all three directories.
IMHO there should be two PRs, one updating `semver` in all three directories and one updating `express` in all three directories.
Otherwise it gets almost impossible to e.g. ignore an update of a dependency (let's say I only want to update `semver` but not `express`).

**Comment by cpswan** [comment]
> Wouldn't that mean, that you no longer get multiple PRs for different dependencies? E.g. if a repositories contains three directories, each with an `npm` project, and each project having e.g. `semver` and `express` as dependencies. When using `groups` as shown above, this would mean, that there's one big PR containing updates to both `semver` and `express` for all three directories. IMHO there should be two PRs, one updating `semver` in all three directories and one updating `express` in all three directories. Otherwise it gets almost impossible to e.g. ignore an update of a dependency (let's say I only want to update `semver` but not `express`).

@ffried you don't have to use a `*` pattern. You could define different groups for different dependencies. But there's probably an anti-pattern lurking here where it becomes necessary (for some) to have a huge number of groups (on a per dependency basis).

**Comment by ffried** [comment]
> You could define different groups for different dependencies. But there's probably an anti-pattern lurking here where it becomes necessary (for some) to have a huge number of groups (on a per dependency basis).

@cpswan The anti-pattern starts a bit earlier for me. Having to use `groups` to get Dependabot to do what it normally does by itself (for single `directory` setups) sounds wrong to me.

**Comment by dhensby** [comment]
I'm struggling to get this to work as (I) expected with terraform.

Here's an example of my config:

```yml
  - package-ecosystem: "terraform"
    directories:
      - "/env/prod"
      - "/env/test"
      - "/env/dev"
    schedule:
      interval: "weekly"
# I have also tried adding the group as described in previous comments
#    groups:
#      terraform:
#        patterns:
#          - "*"
```

However, I only get updates for the first listed directory (`/env/prod`) instead of updates to all the listed directories.

**Comment by carlincherry** [comment]
There's a world where a user sets up multi-directory configuration but also wants the unique PRs vs. grouped PRs generated; that's part of why we opted for the explicitness of requiring a defined group independently of single directory or multiple-directories configuration. I think we could do a better job documenting the expected behavior though, and I have a PR out to our docs team to get this improvement incorporated into the Dependabot docs

**Comment by jan-xyz** [comment]
I've been following along and waiting for the official release and can attest that I would have the use case for separate PRs.

**Comment by lucacome** [comment]
I'm not sure I understand the use case of having multi-directory config but wanting separate PRs, but it seems counterintuitive. Wouldn't it make more sense to have different entries to keep things separate then?

To me this feature is more similar to how the updates for GitHub Actions work now, it doesn't matter how many files I have, the same dependency gets updated in one PR across all the files.

I wouldn't want all the updates in the directory list to be grouped together using `groups`, I just want a PR updating one dependency at a time no matter how many files it's in.

**Comment by Snailedlt** [comment]
@lucacome That's interesting. I would definitely prefer updating one dependency at a time for solo projects or projects where the team is small. But if I'm working on a repo where multiple teams have ownership over different parts of the code (this is often the case in microservice repos) I absolutely don't want to touch others' code, since then I'll have to get approval from everyone before it can get merged.

I'd say there are usecases for both. 1 is more common, but 2 is more important to get right.
The best solution imo would be to add an option to dependabot. Maybe a new "updateStrategy" attribute with the following possible values: "multiFile", "singleFile", where "singleFile" is the default behaviour we have now

**Comment by ipimentelmoraes** [comment]
Hello! We've been testing the multi-directories feature, but we've noticed that, even when using `groups`, dependabot opened duplicated PRs. First, it opened a grouped PR for multiple directories/dependencies (as expected) and, some hours later, separated PRs for dependencies only of the **first** directory of the list.

For reference, here's our `dependabot.yml`:

```yml
version: 2
updates:
  - package-ecosystem: "npm"
    directories:
      - "actions/"
      - "infra/"
    schedule:
      interval: "daily"
    versioning-strategy: increase
    open-pull-requests-limit: 15 
    allow:
      - dependency-name: "collie"
      - dependency-name: "zoo"
    groups:
      version-updates:
        applies-to: version-updates
        patterns: 
          - "*"
```
And the PRs opened:

<img width="1160" alt="image" src="https://github.com/dependabot/dependabot-core/assets/15351744/8969fb46-145b-452b-854b-5985f977997d">

<img width="844" alt="image" src="https://github.com/dependabot/dependabot-core/assets/15351744/6b8623ff-cb32-484a-9553-eb55dee9d04d">

Are we missing anything? We would appreciate some pointers!

**Comment by carlincherry** [comment]
We're getting close to wildcard functionality! We have a question about behavior you'd like or expect to see when you have overlapping directories. Give us your thoughts in this straw poll! https://github.com/dependabot/dependabot-core/discussions/9823

Given an _existing_ configuration file like 👇 

```yaml
version: 2
updates:
  - package-ecosystem: "npm"
    directory: "/"
    schedule:
      interval: "monthly"
    reviewers: ["@person1"]
```

And a user _adds_ the following entry 👇 

```yaml
  - package-ecosystem: "npm"
    directories:
      - "**/*"
    schedule:
      interval: "daily"
    reviewers: ["@person2"]
```

Answer in the poll! https://github.com/dependabot/dependabot-core/discussions/9823

**Comment by dhensby** [comment]
> There's a world where a user sets up multi-directory configuration but also wants the unique PRs vs. grouped PRs generated; that's part of why we opted for the explicitness of requiring a defined group independently of single directory or multiple-directories configuration. I think we could do a better job documenting the expected behavior though, and I have a PR out to our docs team to get this improvement incorporated into the Dependabot docs

@carlincherry Do you have a link to the docs as the only thing I can find about the `directories` option is in [the blog post from April](https://github.blog/changelog/2024-04-29-dependabot-multi-directory-configuration-public-beta-now-available/).

Also, the example in the blog post is:

```yml
version: 2
updates:
  - package-ecosystem: "bundler"
    directories:
      - "/frontend"
      - "/backend"
      - "/admin"
    schedule:
      interval: "weekly"
```

But that results in the behaviour I mentioned above, where dependabot only opens PRs for the first directory and does not try to update the other directories at all. I don't entirely understand why a `groups` definition is required to get it working in all directories, would you mind elaborating on that (or maybe I have the wrong end of the stick when you say "that's part of why we opted for the explicitness of requiring a defined group")?

Just as @lucacome pointed out, I'd really expect this to work like the `github-actions` ecosystem updates, where one dependency that can be updated in all the directories would be grouped into a single PR (assuming no `groups` are configured) - otherwise the `directories` option isn't adding any functionality beyond what already exists when using `directory` (other than saving a few lines of config, of course).

For me, the benefit is that we can reduce the number of PRs dependabot makes when the same dependency needs updating in different directories and thus far less likely to hit the open PR limit. If we have to use groups, then we have to create a definition for each individual dependency in config, which isn't a particularly scalable or maintainable solution as it could mean creating hundreds of groups to get that behaviour (and it would need updating every time a new dependency was added). This seems counter to the stated objective in the blog post:

> Previously, developers with multiple package manifests for the same ecosystem (e.g. npm, pip, gradle) across multiple directories had to create separate `dependabot.yml` configurations for each of those directories. This could lead to many duplicated configurations, and **high maintenance costs if a developer wished to make a change that spanned multiple directories**.

**Comment by carlincherry** [comment]
Thank you all for your feedback and patience during this development process. We're going to launch a private beta for wildcard (globbing) support within the next few weeks! We're soliciting customers interested in joining a private beta for this feature. Please edit this comment and add your repos to the below list if you're interested in the private beta. We anticipate a relatively short beta before we GA.

### Repos for Globbing / Wildcard Private Beta
```
- dvag/workflow-center
- dvag/vp-digital-abschluss-frontend
- dvag/patchreportingv2
``` 

cc @abdulapopoola

**Comment by deki** [comment]
Hmm not everyone has permission to edit the comment and probably you don't want to disclose private repo names.

**Comment by RDhar** [comment]
Thanks @carlincherry, I'd like to submit (my) [DevSecTop/TF-via-PR](https://github.com/DevSecTop/TF-via-PR/blob/main/.github/dependabot.yaml) public repp for testing, which is already trialling the public-beta of the wildcard in dependabot.yml with `package-ecosystem: terraform`.

**Comment by carlincherry** [comment]
@deki yes, please don't disclose anything confidential in this public forum! 

- If you want to add a repo but don't have edit access to the original comment, feel free to comment below and I'll add it to the list
- If you want to add a repo but don't want to comment its name here, feel free to email me and I can add it: carlincherry@github.com

**Comment by v1v** [comment]
I want to trial the wildcards in https://github.com/elastic/apm-pipeline-library/ for the `package-ecosystem: github-actions`. Thanks

**Comment by jorgepiloto** [comment]
Hi @carlincherry, we have a repository of composite actions living in https://github.com/ansys/actions. Could our project be part of the beta testers? Thanks!

**Comment by patatepartie** [comment]
Thanks to the dependabot team for this change ❤️ . It really made updating dependencies easier for me.

I'm not sure if this is the place to report bugs on the new `directories` option (if not, please let me know where it is), but I've run into a quality of life issue: when there is the same dependency being updated in a lot of directories, the section for the dependency is repeated for each directory, leading, in my case, to a truncated output, preventing seeing the changelog for other dependencies.
This happens for me with terraform providers, in a repository where I have many terraform workspaces using the same provider (AWS):
 
<img width="796" alt="Screenshot 2024-06-06 at 9 18 41" src="https://github.com/dependabot/dependabot-core/assets/315939/33a68b61-7dfa-471f-aae5-4c735da31cd0">

That section about the AWS provider is repeated a few time time, then the output is truncated (the markdown ends with `_Description has been truncated_`).

**Comment by billinghamj** [comment]
Could wearemojo/mojo be added please? :)

**Comment by billinghamj** [comment]
@carlincherry We do seem to be having a bit of an issue 🤔

https://github.com/wearemojo/mojo/runs/26128472666

> `The property '#/updates/4/directory' must not include a glob pattern`

Is there a delay in the enabling process?

**Comment by Nishnha** [comment]
Hi folks, this feature is now live! https://github.blog/changelog/2024-06-25-simplified-dependabot-yml-configuration-with-multi-directory-key-directories-and-wildcard-glob-support/

**Comment by deki** [comment]
Wonder how this has been tested. Still facing the issue that only a PR for the first project with a dependency update is being created and all other projects are ignored and kept outdated (both Maven and Gradle). This was pointed out earlier during the beta phase and still occurs today.

**Comment by norrisjeremy** [comment]
> Wonder how this has been tested. Still facing the issue that only a PR for the first project with a dependency update is being created and all other projects are ignored and kept outdated (both Maven and Gradle). This was pointed out earlier during the beta phase and still occurs today.

Yes, seeing the same issue here as well.

**Comment by mattjohnsonpint** [comment]
Same with npm.  Only the first dir in the `directories` list gets updates.

**Comment by JCradock** [comment]
Adding a [grouping](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#groups) fixed the issue of only the first directory being updated using wildcards for us.

**Comment by powerman** [comment]
What will be a minimal working fix? Just a single empty `groups:` key or it needs some content?

**Comment by lucacome** [comment]
@norrisjeremy @powerman I tried grouping and still doesn't work, see #10090

**Comment by deki** [comment]
Just opened another issue related to duplicate PRs in Java https://github.com/dependabot/dependabot-core/issues/10415

**Comment by mattjohnsonpint** [comment]
Not sure what changed, but recently it started working for me where it didn't before.

**Comment by abdulapopoola** [comment]
@mattjohnsonpint , @jakecoffman submitted a couple of fixes a few weeks ago.

**Comment by Dreamsorcerer** [comment]
I've just got round to trying this out and it seems to behave rather weirdly.
I get a PR to update all requirements files (which is what I expect), but also several PRs to update each requirements file individually (not expected, and redundant).
https://github.com/aio-libs/aiohttp-demos/pull/1565
https://github.com/aio-libs/aiohttp-demos/pull/1573
https://github.com/aio-libs/aiohttp-demos/pull/1575

**Comment by Dreamsorcerer** [comment]
As a result, 1 dependency bump results in 2-10 PRs, where 1 is expected, which also causes it go vastly over the open PR limit, which is set to 10 (it's currently managed to open atleast 30..).
